### PR TITLE
@ #57 | should fix rsync auto error

### PR DESCRIPTION
**Pull request by hieptranquoc**
connect https://github.com/teracyhq-incubator/teracy-dev-essential/pull/56.
please check this PR! Thank you so much.

### CODE REVIEW

**Comment by hieptranquoc** [comment]
@hoatle 
Step to reproduce:
- ensure teracy-dev, teracy-dev-entry up to date.
- create `workspace/teracy-dev-entry/config_override.yaml` with content:
```
teracy-dev:
  location:
    sync: true

  entry_location:
    git:
      branch: develop
    sync: false

  extensions:
    - _id: "entry-certs"
      enabled: false

    - _id: "entry-k8s"
      enabled: false

nodes:
  - _id: "0"
    vm:
      synced_folders:
        - _id: "entry-workspace"
          type: rsync
          host: "./workspace"
          guest: "/home/vagrant/workspace"
          rsync__exclude:
            - ".git"
            - ".idea"
            - ".#*"
          create: true

vagrant:
  plugins:
    - _id: "entry-gatling-rsync"
      name: vagrant-gatling-rsync
      version: ""
      state: "installed"
      enabled: true
      # local: false
      env_local: false
      config_key: gatling
      options:
        latency: 0.5
        time_format: "%%H:%%M:%%S"
        rsync_on_startup: true
```
- vagrant up

**Comment by hieptranquoc** [comment]
@hoatle 
Error occurred because rsync__exclude broken. 
https://github.com/hashicorp/vagrant/issues/10974
https://github.com/hashicorp/vagrant/issues/10966
https://github.com/hashicorp/vagrant/issues/11229

**Comment by hoatle** [comment]
@hieptranquoc 
1. **my env**: macOS 10.15.1, vagrant 2.2.4
2. **steps**:
+ clone teracy-dev with the develop branch
+ create teracy-dev/workspace/teracy-dev-entry/config_default.yaml with the following content:
```yaml

nodes:
  - _id: "0"
    vm:
      synced_folders:
        - _id: "entry-workspace"
          type: rsync
          host: "./workspace"
          guest: "/home/vagrant/workspace"
          rsync__exclude:
            - ".git"
            - ".idea"
            - ".#*"
          create: true

vagrant:
  plugins:
    - _id: "entry-gatling-rsync"
      name: vagrant-gatling-rsync
      version: ""
      state: "installed"
      enabled: true
      # local: false
      env_local: false
      config_key: gatling
      options:
        latency: 0.5
        time_format: "%%H:%%M:%%S"
        rsync_on_startup: true
```
+ vagrant up

=> actual: no problem
```
$ vagrant up
==> vagrant: A new version of Vagrant is available: 2.2.6 (installed version: 2.2.4)!
==> vagrant: To upgrade visit: https://www.vagrantup.com/downloads.html

Bringing machine 'node-01' up with 'virtualbox' provider...
==> node-01: Importing base box 'bento/ubuntu-18.04'...
==> node-01: Matching MAC address for NAT networking...
==> node-01: Checking if box 'bento/ubuntu-18.04' version '201910.20.0' is up to date...
==> node-01: A newer version of the box 'bento/ubuntu-18.04' for provider 'virtualbox' is
==> node-01: available! You currently have version '201910.20.0'. The latest is version
==> node-01: '201912.14.0'. Run `vagrant box update` to update.
==> node-01: Setting the name of the VM: teracy-dev_node-01_1576739557750_26819
==> node-01: Clearing any previously set network interfaces...
==> node-01: Preparing network interfaces based on configuration...
    node-01: Adapter 1: nat
==> node-01: Forwarding ports...
    node-01: 22 (guest) => 2222 (host) (adapter 1)
==> node-01: Running 'pre-boot' VM customizations...
==> node-01: Booting VM...
==> node-01: Waiting for machine to boot. This may take a few minutes...
    node-01: SSH address: 127.0.0.1:2222
    node-01: SSH username: vagrant
    node-01: SSH auth method: private key
    node-01: Warning: Connection reset. Retrying...
    node-01: Warning: Remote connection disconnect. Retrying...
    node-01: 
    node-01: Vagrant insecure key detected. Vagrant will automatically replace
    node-01: this with a newly generated keypair for better security.
    node-01: 
    node-01: Inserting generated public key within guest...
    node-01: Removing insecure key from the guest if it's present...
    node-01: Key inserted! Disconnecting and reconnecting using new SSH key...
==> node-01: Machine booted and ready!
==> node-01: Checking for guest additions in VM...
==> node-01: Setting hostname...
==> node-01: Rsyncing folder: /Users/hoatle/teracy-dev/workspace/teracy-dev/workspace/ => /home/vagrant/workspace
==> node-01:   - Exclude: [".vagrant/", ".git", ".idea", ".#*"]
==> node-01: Mounting shared folders...
    node-01: /vagrant => /Users/hoatle/teracy-dev/workspace/teracy-dev
==> node-01: vagrant-gatling-rsync is starting the sync engine because you have at least one rsync folder. To disable this behavior, set `config.gatling.rsync_on_startup = false` in your Vagrantfile.
==> node-01: Doing an initial rsync...
==> node-01: Rsyncing folder: /Users/hoatle/teracy-dev/workspace/teracy-dev/workspace/ => /home/vagrant/workspace
==> node-01:   - Exclude: [".vagrant/", ".git", ".idea", ".#*"]
==> node-01: Watching: /Users/hoatle/teracy-dev/workspace/teracy-dev/workspace
```

**Comment by hieptranquoc** [comment]
@hoatle 
-my env: macOS 10.15.2, vagrant 2.2.6
-steps:
clone teracy-dev with the `v0.6.0-a8` branch
create teracy-dev/workspace/teracy-dev-entry/config_default.yaml with the following content:
```
nodes:
  - _id: "0"
    vm:
      synced_folders:
        - _id: "entry-workspace"
          type: rsync
          host: "./workspace"
          guest: "/home/vagrant/workspace"
          rsync__exclude:
            - ".git"
            - ".idea"
            - ".#*"
          create: true

vagrant:
  plugins:
    - _id: "entry-gatling-rsync"
      name: vagrant-gatling-rsync
      version: ""
      state: "installed"
      enabled: true
      # local: false
      env_local: false
      config_key: gatling
      options:
        latency: 0.5
        time_format: "%%H:%%M:%%S"
        rsync_on_startup: true
```
- vagrant up:
=> actual: error
```
Hieps-MacBook-Pro:k8s-dev hiep$ vagrant up
Bringing machine 'k8s-01' up with 'virtualbox' provider...
==> k8s-01: Importing base box 'bento/ubuntu-18.04'...
==> k8s-01: Matching MAC address for NAT networking...
==> k8s-01: Checking if box 'bento/ubuntu-18.04' version '201906.18.0' is up to date...
==> k8s-01: A newer version of the box 'bento/ubuntu-18.04' for provider 'virtualbox' is
==> k8s-01: available! You currently have version '201906.18.0'. The latest is version
==> k8s-01: '201912.14.0'. Run `vagrant box update` to update.
==> k8s-01: Setting the name of the VM: k8s-dev_k8s-01_1576740217452_71708
==> k8s-01: Clearing any previously set network interfaces...
==> k8s-01: Preparing network interfaces based on configuration...
    k8s-01: Adapter 1: nat
==> k8s-01: Forwarding ports...
    k8s-01: 22 (guest) => 2222 (host) (adapter 1)
==> k8s-01: Running 'pre-boot' VM customizations...
==> k8s-01: Booting VM...
==> k8s-01: Waiting for machine to boot. This may take a few minutes...
    k8s-01: SSH address: 127.0.0.1:2222
    k8s-01: SSH username: vagrant
    k8s-01: SSH auth method: private key
    k8s-01: 
    k8s-01: Vagrant insecure key detected. Vagrant will automatically replace
    k8s-01: this with a newly generated keypair for better security.
    k8s-01: 
    k8s-01: Inserting generated public key within guest...
    k8s-01: Removing insecure key from the guest if it's present...
    k8s-01: Key inserted! Disconnecting and reconnecting using new SSH key...
==> k8s-01: Machine booted and ready!
==> k8s-01: Checking for guest additions in VM...
==> k8s-01: Setting hostname...
==> k8s-01: Rsyncing folder: /Users/hiep/k8s-dev/workspace/ => /home/vagrant/workspace
==> k8s-01:   - Exclude: [".vagrant/", ".git", ".idea", ".#*"]
==> k8s-01: Mounting shared folders...
    k8s-01: /vagrant => /Users/hiep/k8s-dev
==> k8s-01: [vagrant-hostmanager:guests] Updating hosts file on active guest virtual machines...
==> k8s-01: [vagrant-hostmanager:host] Updating hosts file on your workstation (password may be required)...
==> k8s-01: Running provisioner: shell...
    k8s-01: Running: script: Display IP
    k8s-01: IP Address: 10.0.2.15
==> k8s-01: Running provisioner: guest-hosts-fixer (shell)...
    k8s-01: Running: inline script
==> k8s-01: Running action triggers after up ...
==> k8s-01: Running trigger...
==> k8s-01: Doing an initial rsync...
==> k8s-01: Rsyncing folder: /Users/hiep/k8s-dev/workspace/ => /home/vagrant/workspace
==> k8s-01:   - Exclude: [".vagrant/", ".git", ".idea", ".#*"]
[TeracyDevEssential::Config::RsyncRecovery][WARN]: gatling-rsync-auto crashed, retrying...

==> k8s-01: Doing an initial rsync...
==> k8s-01: Rsyncing folder: /Users/hiep/k8s-dev/workspace/ => /home/vagrant/workspace
==> k8s-01:   - Exclude: [".vagrant/", ".git", ".idea", ".#*"]
[TeracyDevEssential::Config::RsyncRecovery][WARN]: gatling-rsync-auto crashed, retrying...

==> k8s-01: Doing an initial rsync...
==> k8s-01: Rsyncing folder: /Users/hiep/k8s-dev/workspace/ => /home/vagrant/workspace
==> k8s-01:   - Exclude: [".vagrant/", ".git", ".idea", ".#*"]
[TeracyDevEssential::Config::RsyncRecovery][WARN]: gatling-rsync-auto crashed, retrying...

==> k8s-01: Doing an initial rsync...
==> k8s-01: Rsyncing folder: /Users/hiep/k8s-dev/workspace/ => /home/vagrant/workspace
==> k8s-01:   - Exclude: [".vagrant/", ".git", ".idea", ".#*"]
[TeracyDevEssential::Config::RsyncRecovery][WARN]: gatling-rsync-auto crashed, retrying...

==> k8s-01: Doing an initial rsync...
==> k8s-01: Rsyncing folder: /Users/hiep/k8s-dev/workspace/ => /home/vagrant/workspace
==> k8s-01:   - Exclude: [".vagrant/", ".git", ".idea", ".#*"]
[TeracyDevEssential::Config::RsyncRecovery][WARN]: gatling-rsync-auto crashed, retrying...

==> k8s-01: Doing an initial rsync...
==> k8s-01: Rsyncing folder: /Users/hiep/k8s-dev/workspace/ => /home/vagrant/workspace
==> k8s-01:   - Exclude: [".vagrant/", ".git", ".idea", ".#*"]
[TeracyDevEssential::Config::RsyncRecovery][WARN]: gatling-rsync-auto crashed, retrying...

==> k8s-01: Doing an initial rsync...
==> k8s-01: Rsyncing folder: /Users/hiep/k8s-dev/workspace/ => /home/vagrant/workspace
==> k8s-01:   - Exclude: [".vagrant/", ".git", ".idea", ".#*"]
^C[TeracyDevEssential::Config::RsyncRecovery][WARN]: gatling-rsync-auto crashed, retrying...

Exiting due to interrupt.
```

File: lib/teracy-dev-essential/config/rsync_recovery.rb [hunk 1]
```diff
@@ -47,6 +47,10 @@ def configure_node(settings, config)
           end
         end
 
+        if TeracyDev::Util.require_version_valid?(@vagrant_version, ">= 2.2.4") and TeracyDev::Util.require_version_valid?(@vagrant_version, "< 2.2.7")
```
Comment (hoatle) [inline lib/teracy-dev-essential/config/rsync_recovery.rb]: we force to use rsync-auto on vagrant 2.2.5 and 2.2.6, so use `>= 2.2.5` and `< 2.2.7` instead

Comment (hoatle) [reply lib/teracy-dev-essential/config/rsync_recovery.rb]: add logger.info
```ruby
@logger.info("gatling-rsync-auto does not work on vagrant 2.2.5 and 2.2.6, so going to use rsync-auto instead")
```

**Review by hoatle — CHANGES_REQUESTED** [review]
@hieptranquoc please check my comment.

**Comment by hieptranquoc** [comment]
I applied all the comments! Please check this PR again! Thank you!

**Review by hoatle — APPROVED** [review]
lgtm

**Pull request merged**